src/device/wg_controller.rs

```diff
@@ -236,7 +255,31 @@ async fn consolidate_wg_peers<
         }
 
         let is_actual_peer_proxying = is_peer_proxying(actual_peer, &proxy_endpoints);
-        if is_actual_peer_proxying && !is_peer_proxying(&requested_peer.peer, &proxy_endpoints) {
+        let is_reqested_peer_proxying = is_peer_proxying(&requested_peer.peer, &proxy_endpoints);
+
+        let event = AnalyticsEvent {
+            public_key: *key,
+            dual_ip_addresses: actual_peer.get_dual_ip_addresses(),
+            tx_bytes: actual_peer.tx_bytes.unwrap_or_default(),
+            rx_bytes: actual_peer.rx_bytes.unwrap_or_default(),
+            peer_state: actual_peer.state(),
+            timestamp: Instant::now(),
+        };
+        if is_actual_peer_proxying {
+            aggregator.report_peer_state_relayed(&event).await;
+        } else if let Some(us) = upgrade_sync {
+            if let Some(accepted_direct_session) = us.get_accepted_session(*key).await {
```

Comment:
:thinking: I am afraid there might be edge-case here.

Lets imagine we have accepted some upgrade, and upgraded connection to direct. Later, for some reason, another upgrade sync request is sent, it gets accepted, but because we are already in direct, no endpoint change in wireguard will occur. This would report new upgrade sync info, but keep old wireguard endpoint for, potentially, indefinite amount of time.

In order to fix it, one possible approach would be to expand accepted_session with LRU cache, which would hold some bounded amount of upgrade sync sessions, keyed by endpoint (`IP:port`). In here we would get accepted session by pubic key as well as currently "installed" endpoint in wireguard. If no such session could be found, we can report "Unknown:Unknown" as endpoint provider (**Note:** [this](https://github.com/NordSecurity/libtelio/blob/v5.0.0-rc4/crates/telio-nurse/src/aggregator.rs#L45) would have to be expanded with "Unknown" case.)

Reply:
oh

Actually I am wrong. As new upgrade request would get honored even if we are already in direct. And in practice, if we would receive upgrade syn request, and are "planning" to ignore it, it is probably better to reject it instead of accepting it.

Ok, so this edge-case turned out to be not real.